File: backend/packages/nestjs/test/smoke.ts
/**
 * End-to-end smoke test.
 * Run with:  npx ts-node test/smoke.ts
 *
 * Verifies:
 * 1. AST parser produces a valid graph from the fixture directory.
 * 2. Sidecar server starts and serves the graph on the configured port.
 * 3. Health endpoint returns 200.
 * 4. Graph endpoint returns the parsed nodes.
 */

import * as http from 'http';
import * as path from 'path';

import { AstParserService } from '../src/ast/ast-parser.service';
import { CacheService } from '../src/cache/cache.service';
import { SidecarService } from '../src/sidecar/sidecar.service';

const SIDECAR_PORT = 4568; // Use a different port so it doesn't conflict with production
const FIXTURE_DIR = path.resolve(__dirname, 'fixtures');

function get(url: string): Promise<{ status: number; body: string }> {
  return new Promise((resolve, reject) => {
    http.get(url, (res) => {
      let body = '';
      res.on('data', (chunk) => (body += chunk));
      res.on('end', () => resolve({ status: res.statusCode ?? 0, body }));
    }).on('error', reject);
  });
}

async function run(): Promise<void> {
  console.log('--- FlowMapper smoke test ---\n');

  // 1. Parse fixtures
  const parser = new AstParserService();
  const graph = parser.parse(FIXTURE_DIR);
  console.log(`[AST]  Parsed ${graph.nodes.length} nodes, ${graph.edges.length} edges`);

  const taggedNode = graph.nodes.find((n) => n.customTag);
  if (!taggedNode) throw new Error('Expected at least one node with a @FlowStep tag');
  console.log(`[AST]  @FlowStep node → label: "${taggedNode.label}"`);

  const docstringNode = graph.nodes.find((n) => n.docstring);
  if (!docstringNode) throw new Error('Expected at least one node with a JSDoc docstring');
  console.log(`[AST]  JSDoc node → docstring: "${docstringNode.docstring?.slice(0, 60)}..."`);

  // 2. Cache round-trip
  const cache = new CacheService('/tmp/flow-map-smoke-cache');
  const hash = cache.hashBody('function test() {}');
  cache.set('smoke::test', hash, 'Test summary');
  const hit = cache.get('smoke::test', hash);
  if (hit !== 'Test summary') throw new Error(`Cache miss after set: got "${hit}"`);
  console.log('[Cache] Set/get round-trip OK');

  const miss = cache.get('smoke::test', 'different-hash');
  if (miss !== undefined) throw new Error('Expected undefined on hash mismatch');
  console.log('[Cache] Hash-mismatch returns undefined OK');

  // 3. Sidecar start + HTTP verification
  const sidecar = new SidecarService();
  sidecar.updateGraph(graph);
  await sidecar.start(SIDECAR_PORT);
  console.log(`[Sidecar] Started on http://localhost:${SIDECAR_PORT}`);

  const health = await get(`http://localhost:${SIDECAR_PORT}/api/flow-map/health`);
  if (health.status !== 200) throw new Error(`Health returned ${health.status}`);
  console.log('[Sidecar] /health → 200 OK');

  const graphResp = await get(`http://localhost:${SIDECAR_PORT}/api/flow-map/graph`);
  if (graphResp.status !== 200) throw new Error(`/graph returned ${graphResp.status}`);
  const parsed = JSON.parse(graphResp.body);
  if (parsed.data.nodes.length !== graph.nodes.length) {
    throw new Error(`Expected ${graph.nodes.length} nodes in API response, got ${parsed.data.nodes.length}`);
  }
  console.log(`[Sidecar] /graph → 200, ${parsed.data.nodes.length} nodes, ${parsed.data.edges.length} edges`);

  await sidecar.stop();
  console.log('\n✓ All smoke checks passed');
}

run().catch((err) => {
  console.error('\n✗ Smoke test failed:', err.message);
  process.exit(1);
});

File: backend/packages/python/test/smoke.py
"""
End-to-end smoke test.
Run with:  python test/smoke.py   (from packages/python/)

Verifies:
1. AST parser produces a valid graph from the fixture directory.
2. Cache set/get round-trip works correctly.
3. Sidecar server starts and serves the graph.
4. Health and graph endpoints return expected responses.
"""

import asyncio
import json
import os
import sys
import tempfile
import time
import urllib.request

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from flow_map.ast_parser import AstParserService
from flow_map.cache import CacheService
from flow_map.sidecar import SidecarService

SIDECAR_PORT = 4569  # Use a distinct port to avoid conflicts
FIXTURE_DIR = os.path.join(os.path.dirname(__file__), "fixtures")


def get(url: str, timeout: float = 5.0) -> tuple[int, dict]:
    with urllib.request.urlopen(url, timeout=timeout) as resp:
        return resp.status, json.loads(resp.read())


def wait_for_server(port: int, retries: int = 10, delay: float = 0.3) -> None:
    for _ in range(retries):
        try:
            get(f"http://localhost:{port}/api/flow-map/health")
            return
        except Exception:
            time.sleep(delay)
    raise RuntimeError(f"Sidecar did not start on port {port} within {retries * delay}s")


def run() -> None:
    print("--- FlowMap smoke test ---\n")

    # 1. Parse fixtures
    parser = AstParserService()
    graph = parser.parse(FIXTURE_DIR)
    print(f"[AST]  Parsed {len(graph.nodes)} nodes, {len(graph.edges)} edges")

    tagged = next((n for n in graph.nodes if n.custom_tag), None)
    if not tagged:
        raise AssertionError("Expected at least one node with a @flow_step tag")
    print(f'[AST]  @flow_step node → label: "{tagged.label}"')

    docstring_node = next((n for n in graph.nodes if n.docstring), None)
    if not docstring_node:
        raise AssertionError("Expected at least one node with a docstring")
    print(f'[AST]  Docstring node → "{(docstring_node.docstring or "")[:60]}..."')

    # 2. Cache round-trip
    with tempfile.TemporaryDirectory() as tmp:
        cache = CacheService(tmp)
        h = cache.hash_body("def test(): pass")
        cache.set("smoke::test", h, "Test summary")

        hit = cache.get("smoke::test", h)
        assert hit == "Test summary", f"Cache hit returned wrong value: {hit!r}"
        print("[Cache] Set/get round-trip OK")

        miss = cache.get("smoke::test", "wrong-hash")
        assert miss is None, f"Expected None on hash mismatch, got: {miss!r}"
        print("[Cache] Hash-mismatch returns None OK")

    # 3. Sidecar server
    sidecar = SidecarService()

    import dataclasses
    graph_dict = {
        "nodes": [dataclasses.asdict(n) for n in graph.nodes],
        "edges": [{"from": e.from_id, "to": e.to_id} for e in graph.edges],
        "generatedAt": graph.generated_at,
    }
    sidecar.update_graph(graph_dict)
    sidecar.start(SIDECAR_PORT)
    wait_for_server(SIDECAR_PORT)
    print(f"[Sidecar] Started on http://localhost:{SIDECAR_PORT}")

    status, body = get(f"http://localhost:{SIDECAR_PORT}/api/flow-map/health")
    assert status == 200, f"Health returned {status}"
    print("[Sidecar] /health → 200 OK")

    status, body = get(f"http://localhost:{SIDECAR_PORT}/api/flow-map/graph")
    assert status == 200, f"/graph returned {status}"
    assert len(body["data"]["nodes"]) == len(graph.nodes), (
        f"Expected {len(graph.nodes)} nodes, got {len(body['data']['nodes'])}"
    )
    print(
        f"[Sidecar] /graph → 200, "
        f"{len(body['data']['nodes'])} nodes, {len(body['data']['edges'])} edges"
    )

    sidecar.stop()
    print("\n✓ All smoke checks passed")


if __name__ == "__main__":
    try:
        run()
    except Exception as err:
        print(f"\n✗ Smoke test failed: {err}", file=sys.stderr)
        sys.exit(1)
